Use UUID's to generate random tempname on Windows

musm opened the pull request:
cc @StefanKarpinski @vtjnash  @davidanthoff  

I wrote up this quick implementation that use UIUD's to generate random file names as recommended in the MSFT docs. Note, we already load the library `"Rpcrt4.dll"` so it shouldn't be an issue to use it here. (According to `Libdl.dll()` this library is already loaded on Windows when using Julia. )

musm committed:
Use UIUD to create random tempname on Windows

musm commented on base/file.jl:
```diff
@@ -510,22 +510,47 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
+let
+global tempname
+
+struct GUID
```
I want to hide the `GUID`  struct definition using a let block, but apparently they escape  let scoping ?  (noted by `@mbauman ` )

StefanKarpinski replied:
We don't have local named types so I think you can just get rid of the `let` wrapper.

musm commented on base/file.jl:
```diff
@@ -510,22 +510,47 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
+    @assert ccall((:RpcStringFreeW, :Rpcrt4), Cint, (Ref{Ptr{Cwchar_t}},), nameptr) == 0
+
+    filename = joinpath(parent, name)
+    if !ispath(filename)
```
Are we ok with this. I feel like this should just error (maybe?), since it will be an incredible rare scenario. In the previous code, we did keep looping until a unique name was generated. Here,  in the unlikely scenario it is a path, we recursively call the function.

davidanthoff replied:
I think maybe if `cleanup` is false we could not call `ispath` at all? It would be nice if this function would actually not touch the file system at all.

I tend to think that just assuming we'll never run into a conflict is ok here?

musm replied:
> I think maybe if cleanup is false we could not call ispath at all?

Agreed, in fact it's not guaranteed to return a unique filename (even if very unlikely)

musm replied:
Since @StefanKarpinski worked on this, could you comment on  ?

> It would be nice if this function would actually not touch the file system at all.

I don't see how it would be possible to do this in the scenario, since if we don't check `!ispath(filename)` ,   `temp_cleanup_later` could accidentally remove the file which we did not generate.

davidanthoff commented on base/file.jl:
```diff
@@ -510,22 +510,47 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
+
+    id = Ref{GUID}()
+    r = ccall((:UuidCreate,:Rpcrt4), Cint, (Ref{GUID},), id)
+    r == 0 || (r == RPC_S_UUID_LOCAL_ONLY && (@warn "UIUD guaranteed to be unique to this computer only"; true)) || (r == RPC_S_UUID_NO_ADDRESS && error("Cannot get Ethernet or token-ring hardware address for this computer."))
```
Do we need the warning? It is not clear when that case would actually ever happen (no network card installed?), but even if it does, it seems that a machine unique filename is good enough and doesn't need to emit a warning?

StefanKarpinski replied:
Seems fine not to emit this warning to me, but I'm not an expert in this platform.

musm replied:
> The UuidCreateSequential function returns RPC_S_UUID_LOCAL_ONLY when the originating computer does not have an ethernet/token ring (IEEE 802.x) address. In this case, the generated UUID is a valid identifier, and is guaranteed to be unique among all UUIDs generated on the computer. However, the possibility exists that another computer without an ethernet/token ring address generated the identical UUID. Therefore you should never use this UUID to identify an object that is not strictly local to your computer. Computers with ethernet/token ring addresses generate UUIDs that are guaranteed to be globally unique.

I am wondering if there are implications for distributed process over a network utilizing this function to generate tempnames. It may be beneficial to emit this warning in the rare case that this is true?

davidanthoff replied:
I think `UuidCreate` is what is used by default to create GUIDS on Windows, so I think I would just stick with that, and treat a return value of `RPC_S_UUID_LOCAL_ONLY` as fine and not worth a warning.

davidanthoff commented on base/file.jl:
```diff
@@ -510,22 +510,47 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
+
+    @assert ccall((:RpcStringFreeW, :Rpcrt4), Cint, (Ref{Ptr{Cwchar_t}},), nameptr) == 0
+
+    filename = joinpath(parent, name)
```
Should we prefix the filename with something like `jl_`? I think there is already a constant defined in the code somewhere. That way one could easily identify the temp files created by Julia.

davidanthoff commented:
Also pinging @JeffBezanson, who wrote the current version of `tempname` for Windows according to blame.

Context is https://github.com/JuliaLang/julia/pull/33741 and https://github.com/JuliaLang/julia/pull/33765.

StefanKarpinski commented:
> Also pinging @JeffBezanson, who wrote the current version of `tempname` for Windows according to blame.

Seems unlikely—probably just an artifact of him being the person to move the code to the current location.

davidanthoff commented:
> Seems unlikely—probably just an artifact of him being the person to move the code to the current location.

It looks to me as if he really wrote it: https://github.com/JuliaLang/julia/commit/5e2ff127f861ead7b08ce782b830069d77c97f2f

musm commented:
we have unrelated failures here

vtjnash commented:
package_win32 is a segfault in tempname

musm commented:
oops your right. Might be a fluke. The previous build both ran into the same  unrelated repl kill test errors

musm commented:
```
temp_prefix = "jl_"

struct GUID
    l1::Culong
    w1::Cushort
    w2::Cushort
    b::NTuple{8,Cuchar}
end

function tempname(parent::AbstractString=tempdir(); cleanup::Bool=true)
    isdir(parent) || throw(ArgumentError("$(repr(parent)) is not a directory"))

    RPC_S_UUID_LOCAL_ONLY = 1824
    RPC_S_UUID_NO_ADDRESS = 1739
    RPC_S_OUT_OF_MEMORY = 14

    id = Ref{Ptr{GUID}}(0)
    r = ccall((:UuidCreateSequential,:Rpcrt4), Clong, (Ref{Ptr{GUID}},), id)
    (r == 0 || r == RPC_S_UUID_LOCAL_ONLY) || (r == RPC_S_UUID_NO_ADDRESS && error("Cannot get Ethernet or token-ring hardware address for this computer."))

    nameptr = Ref{Ptr{Cwchar_t}}(0)
    r = ccall((:UuidToStringW, :Rpcrt4), Cint, (Ref{Ptr{GUID}}, Ref{Ptr{Cwchar_t}}), id, nameptr)
    r == 0 || (r == RPC_S_OUT_OF_MEMORY && error("Out of memory")) || error("cannot convert UIUD to string")

    namebuf = unsafe_wrap(Vector{Cwchar_t}, nameptr[], ccall(:wcslen, UInt, (Ptr{Cwchar_t},), nameptr[]))
    name = transcode(String, namebuf)

    @assert ccall((:RpcStringFreeW, :Rpcrt4), Cint, (Ref{Ptr{Cwchar_t}},), nameptr) == 0

    filename = joinpath(parent, temp_prefix * name)
    # if !cleanup
    #     return filename
    # else
    #     # in the unlikley scenario filename is a path, call this function again regenerate a unique filename
    #     return !ispath(filename) ? (temp_cleanup_later(filename); filename) : tempname(parent; cleanup=cleanup)
    # end
end


@show tempname()

```

reproducible test script. The above is causing a segfault on win32 for some reason. Sometimes I get a misaligned pointer 2 bytes, and sometimes it just segfaults.

vtjnash commented on test/file.jl:
```diff
@@ -50,6 +50,20 @@ end
 using Random
 
+@testset "that temp names are actually unique" begin
+    temps = [tempname(cleanup=false) for _ = 1:1000]
```
This has an expected failure rate on our CI of 0.001% (`1 - e^(-1000^2 / 62^6 / 2)`) which means 1 in 100k CI runs (or about 1 in 10k commits) would be expected to fail this test (over the current lifetime of this project, that would mean there'd have been about 2-3 failures)

davidanthoff replied:
I don't follow. The filenames that are generated in this PR are 128 bit UUIDs, which, as far as I can tell, will lead to a way, way, way lower collision rate than what you compute here.

musm replied:
I think the point that Jameson is making is that this test is unnecessarily strict, which I agree (I only used 1000 to test the PR). Something like 100, is more than sufficient.

staticfloat replied:
At least this does not use Julia's RNG, which gets reset on every `@testset`.

musm commented on base/file.jl:
```diff
@@ -513,19 +513,41 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
+        return filename
+    else
+        # in the unlikley scenario filename is a path, call this function again regenerate a unique filename
+        return !ispath(filename) ? (temp_cleanup_later(filename); filename) : tempname(parent; cleanup=cleanup)
```
I would like to get rid of `ispath` if possible.  Is it ok to replace these last statements with
```julia
cleanup && temp_cleanup_later(filename)
return filename
```
Like in the non-windows version?

musm commented:
Ok 32-bit windows is now fixed (calling convention should've been `stdcall`)

musm commented:
This is now ready for review

musm commented on base/file.jl:
```diff
@@ -513,20 +513,37 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
+
+    ccall((:RpcStringFreeW, :Rpcrt4), stdcall, Cint, (Ref{Ptr{Cwchar_t}},), nameptr) == 0
+
+    filename = joinpath(parent, temp_prefix * name)
```
For reviewers:
Note we prefix the path with `jl_`

examples include 
```
"C:\\Users\\Mus\\AppData\\Local\\Temp\\jl_ed1e9da2-b30b-4354-8149-1b220e275e99"
```

musm commented:
Both Windows buildbot failures are unrelated busybox failures.

musm commented:
Note, this PR does lead to a bigger difference than what we currently have between the linux and windows behavior for `tempname`; I'm not sure if we care about this. On linux we have names like `jl_XXXXXX` and on windows now `jl_c1b1ff4c-c90a-46a9-8348-85ad8170ff04`

musm commented:
I think the `jl_` prefix doesn't really make too much sense here. These are temporary filenames and thus the name shouldn't have meaning, so why bother with the unnecessary prefix here, where we are generating a UUID anyways.  I think it makes more sense  in the linux version with `jl_XXXXXX`, but with the UUID's I'd argue it's not needed.

davidanthoff commented:
I sometimes go into my temp folder to delete stuff, and I find it quite helpful in those cases to be able to tell which file originated from what application, so that I can do select deletes, so I kind of like the `jl_` prefix. Is there any real downside to it?

musm commented:
Ok just wondered if someone had thoughts about that.

This PR from my standpoint is ready. Tests are all passing.

vtjnash commented on test/file.jl:
```diff
@@ -50,6 +50,20 @@ end
+    temps = map(1:100) do _
+        path, io = mktemp(cleanup=false)
+        close(io)
+        return path
```
```suggestion
        rm(path, force=true)
        return path
```

musm replied:
Shouldn't we then delete L63-65...?

musm commented:
All green on the Western front

vtjnash commented on base/file.jl:
```diff
@@ -513,20 +513,37 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
+
+    nameptr = Ref{Ptr{Cwchar_t}}()
+    r = ccall((:UuidToStringW, :Rpcrt4), stdcall, Cint, (Ref{GUID}, Ref{Ptr{Cwchar_t}}), id, nameptr)
+    r == 0 || (r == RPC_S_OUT_OF_MEMORY && error("Out of memory")) || error("cannot convert UIUD to string")
```
```suggestion
    r == 0 || windowserror("UuidToString", r % UInt32) # a RPC_STATUS is just a WINERROR with a different signedness of the type
```

vtjnash commented on base/file.jl:
```diff
@@ -513,20 +513,37 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
+
+    id = Ref{GUID}()
+    r = ccall((:UuidCreate,:Rpcrt4), stdcall, Cint, (Ref{GUID},), id)
+    (r == 0 || r == RPC_S_UUID_LOCAL_ONLY) || (r == RPC_S_UUID_NO_ADDRESS && error("Cannot get Ethernet or token-ring hardware address for this computer."))
```
```suggestion
    if !(r == 0 || r == RPC_S_UUID_LOCAL_ONLY || r == RPC_S_UUID_NO_ADDRESS) # ignore expected errors that we don't care about (and I suspect that since Vista, they aren't possible anymore)
        windowserror("UuidCreate", r % UInt32) # Throw on unexpected errors. Note, a RPC_STATUS is just a WINERROR with a different signedness of the type.
    end
```

vtjnash commented on base/file.jl:
```diff
@@ -513,20 +513,37 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
+    namebuf = unsafe_wrap(Vector{Cwchar_t}, nameptr[], ccall(:wcslen, UInt, (Ptr{Cwchar_t},), nameptr[]))
+    name = transcode(String, namebuf)
+
+    ccall((:RpcStringFreeW, :Rpcrt4), stdcall, Cint, (Ref{Ptr{Cwchar_t}},), nameptr) == 0
```
```suggestion
    ccall((:RpcStringFreeW, :Rpcrt4), stdcall, Cint, (Ref{Ptr{Cwchar_t}},), nameptr) # realistically, this can't fail
```

vtjnash commented on base/file.jl:
```diff
@@ -513,20 +513,37 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
+
+    ccall((:RpcStringFreeW, :Rpcrt4), stdcall, Cint, (Ref{Ptr{Cwchar_t}},), nameptr) == 0
+
+    filename = joinpath(parent, temp_prefix * name)
```
I'm tempted to say we should also do:
```suggestion
    @assert !ispath(filename)
```
to guard against some extreme forms of malfeasance.

Thoughts?

musm replied:
I'm not sure if we want to do this. I prefer not to include this since, ideally, this function should avoid touching the filesystem at all.

vtjnash replied:
It's expected to touch the filesystem in this way on all other platforms though (it's part of the definition of posix tempnam we use otherwise)

musm replied:
What if we just restrict this to the branch where `cleanup==true` ?

vtjnash reviewed (commented):
The overall approach seems good. Some open-ended questions:

1. Is it OK to have filenames with `-` in them? Seems like it should be a pretty well supported character, but if not, we could add `name = filter(!=('-'), name)` (although, for that matter we could also use the existing code in uuid.jl for turning these into strings, as that definition of struct UUID is valid as a drop-in replacement for this struct GUID usage)
2. Is it OK for the filenames to be so long? Since the tempdir may already be pretty deep in the path (`C:\Documents and Settings\Your Full Username Here\AppData\Local\Temp\`), this'll reduce the maximum supported username length from 201 to 174 utf16 code points.
3. Is it OK to use filenames without suffixes?

All of these seem OK to me, but I just wanted to voice them.

staticfloat commented:
> All of these seem OK to me, but I just wanted to voice them.

These all seem OK to me as well.

musm commented:
> 3\. Is it OK to use filenames without suffixes?

It's fine, but maybe we should add a `.tmp` suffix here?

> * Is it OK to have filenames with `-` in them? Seems like it should be a pretty well supported character, but if not, we could add `name = filter(!=('-'), name)` (although, for that matter we could also use the existing code in uuid.jl for turning these into strings, as that definition of struct UUID is valid as a drop-in replacement for this struct GUID usage)

![image](https://user-images.githubusercontent.com/4319522/69642053-306a4680-102f-11ea-87f4-ad9f65dc39d4.png)

I'm not sure which application this is, but you can see the enclose the generated UUID with braces.

It's ok to use filenames with `-` in them.  But I actually think I agree to filter the `-`. I don't think we should use the `uuid.jl`  file to convert to a string, since that file is loaded later in `Base.jl`

musm commented:
@vtjnash  Do you mind taking another look at the updated PR?

I added (1) `.tmp` suffix; (2) filtered the `-` away; (3) refactored the tempname generation into a seperate function; (4) and addressed review comments.

musm committed:
Use underscores and remove extension

StefanKarpinski commented:
Can we crunch the UUID down to 10 or so characters from `0-9a-zA-Z`? I'm worried about Windows systems with 260-byte path length limits, which we've encountered occasionally in Pkg stuff. Adding this many bytes to the temp path name is slightly worrying.

rfourquet commented:
Probably a naive question, but I still can't understand why we don't use the same code for Windows and Unix, with a home-made function, along the lines of:

```julia
_randstring() = String(rand(['a':'z'; 'A':'Z'; '0':'9';], 10)) # randstring() is in `Random`

function tempname2(parent::AbstractString=tempdir(); cleanup::Bool=true)
    while true
        filename = joinpath(parent, temp_prefix * _randstring())
        !ispath(filename) && return filename
    end
end
```

StefanKarpinski commented on base/file.jl:
```diff
@@ -513,20 +513,45 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
+
+    ccall((:RpcStringFreeW, :Rpcrt4), stdcall, Cint, (Ref{Ptr{Cwchar_t}},), nameptr)
+
+    name = replace(name, '-' => '_')
```
```suggestion
    n = UInt64(0)
    for c in name
        d = '0' <= c <= '9' ? c - '0' :
             'a' <= c <= 'f' ? c - 'a' :
             'A' <= c <= 'F' ? c - 'A' : continue
        n = (n << 4) + (d % UInt64)
    end
    slug = StringVector(10)
    chars = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
    for i = 1:length(slug)
        slug[i] = chars[(n % length(chars)) + 1]
        n ÷= length(chars)
    end
    name = String(slug)
```

vtjnash replied:
a-z and A-Z generally map to the same files. should we just use the uppercase letters?

vtjnash replied:
Seems rather convoluted though to convert a UInt128 -> GUID -> String-base16 -> UInt64 -> String-base62. I would think it'd be much shorter to do:
```
id = Ref{UInt128}()
# UuidCreate(id)
name = string(id[] % UInt64, base=36, pad=5) # between 5 and 13 characters, to avoid special file names like COM1
name = name[1:min(end, 10)] # truncate to max of 10 characters (why 10? I have no idea, it came from the random number generator)
```

StefanKarpinski replied:
Seems better to generate both upper and lowercase so the chance of collision is lower on systems with case sensitivity.  Of course, that increases the chance of collision on systems without it. Perhaps sample from a-z, A-Z with half the frequency of 0-9?

musm replied:
@vtjnash  I like the simplicity of your proposal. What does `pad=5` do? 
I can update the PR to this change. It makes the code a lot shorter and simpler.

vtjnash replied:
It guarantees that it can't generate an invalid filename like `COM1`

musm commented:
More generally, do we need this function? It's intended usage is not recommended as it can cause security holes. I think we should emulate Python's approach in this area https://docs.python.org/2/library/tempfile.html

davidanthoff commented:
Just for reference, [here](https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Private.CoreLib/src/System/IO/Path.cs#L253) is how .Net Core implements this. They don't start with a UUID at all, but some random bytes from the crypto API in Windows, and then turn that into a 83 filename...

rfourquet commented:
I don't want to be annoying, but if it would be useful to have a bit of information written somewhere to explain why a very [simpler approach](https://github.com/JuliaLang/julia/pull/33785#issuecomment-564668194) is not used here.

StefanKarpinski commented:
Yes, could someone please address @rfourquet's seemingly reasonable suggestion/question?

JeffBezanson commented:
We could indeed use @rfourquet 's suggestion, we just need to use the system random device RNG. On windows it's just one system call.

StefanKarpinski commented:
Or more directly, just ccall `(:SystemFunction036, :Advapi32)` like we do to implement RandomDevice for Windows:

https://github.com/JuliaLang/julia/blob/fdcaa0674bbd3c2b3f/stdlib/Random/src/RNGs.jl#L46-L47

That avoids needing to surreptitiously reach into the Random stdlib from Base, which we could do, but is not entirely ideal. It looks like this might work:
```
nchars = 10
nbytes = ceil(Int, nchars*(log2(62)/8))
A = Vector{UInt8}(undef, nbytes)
ccall((:SystemFunction036, :Advapi32), stdcall, UInt8, (Ptr{Cvoid}, UInt32), A, sizeof(A))
```
Then turn the bytes in `A` into a ten character slug for the path name.

JeffBezanson commented:
Triage thinks this is a bug fix and so does not block feature freeze.

Keno commented:
Seems like this didn't work on win64?

musm commented:
It was passing before I made the change to  use underscores.
I need to update the PR with the some of the requested changes, which I should get to later today.

StefanKarpinski commented:
I’ll repeat: putting a UUID in the temp name is too long. This should not be longer that than 10 or so chars.

musm commented:
```julia
import Base.StringVector

function _rand_string()
	nchars = 10
	A = Vector{UInt8}(undef, nchars)
	ccall((:SystemFunction036, :Advapi32), stdcall, UInt8, (Ptr{Cvoid}, UInt32), A, sizeof(A))

	slug = StringVector(10)
	chars = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
	for i = 1:nchars
	    slug[i] = chars[(A[i] % length(chars)) + 1]
	end
	return name = String(slug)
end
```

which approach is better to use the above, or the one in the last commit?

musm committed:
Truncate to 10 chars the UUID

StefanKarpinski commented:
The implementation in the comment seems better to me.

musm commented:
Does that actually create a bias?

musm committed:
Generate the random name from a random byte array

musm commented:
Anyways I tend to agree that it is probably a better approach, and is similar in approach to that used in the .Net Core   that @davidanthoff linked.

musm committed:
Update file.jl

StefanKarpinski commented:
> Does that actually create a bias?

Yes: since `divrem(256, 62) == (4, 8)` the first 8 characters (0-7) are more likely than the rest by odds of 5:4 (so 25% more likely). However, that's ok since for maximum distinguishability on a case-insensitive filesystem we should be choosing the digits twice as often as the letters, so the bias actually makes the resulting slug slightly more likely to be distinguishable. I think we should proceed with this anyway.

Keno commented on base/file.jl:
```diff
@@ -517,20 +517,27 @@ function mktemp(parent::AbstractString=tempdir(); cleanup::Bool=true)
+function _rand_string()
+	nchars = 10
+	A = Vector{UInt8}(undef, nchars)
+	ccall((:SystemFunction036, :Advapi32), stdcall, UInt8, (Ptr{Cvoid}, UInt32), A, sizeof(A))
```
This function can fail. The return value should be checked, but there's not much we can do to recover to just throw an error.

Keno replied:
Actually, we do the same thing in the stdlib, so I'm gonna merge this and then fix them both at once.

rfourquet commented:
Given that this implementation is biased, in a non-problematic way, wouldn't if be fine to even use `Libc.rand()` then, and simplify slightly `_rand_string()` ? (not that I suggest doing it, now that it's already merged; just a genuine question).

musm commented:
wouldn't hurt to open a PR @rfourquet if you have something in mind, thanks.
